### ISSUE TITLE
On detach changeDetectorRef.detectChanges and ngOnChanges are out of sync

### ISSUE THREAD

**Issue by daixtrose** [angular/angular#10630]
**I'm submitting a ...** 
[ x ] bug report
[ x ] feature request

**Current behavior**

When using 

``` typescript
constructor(private changeDetectorRef: ChangeDetectorRef) {
    changeDetectorRef.detach();
    setInterval(() => {
        this.changeDetectorRef.detectChanges();
    }, 5000);
  }
```

The change detection takes place every 5 seconds as expected, but `ngOnChanges` is called at a higher rate, which AFAICS means that expensive renderings might still take place at much higher frequency.

**Reproduction of the problem**

See http://plnkr.co/edit/K5Lhcb8N6cXc7NXZdtSB?p=preview 

**What is the expected behavior?**

I had expected that `ngOnChanges` and change detection are in sync

**What is the motivation / use case for changing the behavior?**

The situation IMHO is complex. The context is that  #7054 to my surprise was closed w/o code change and declared _works as intended_. This led me to some further investigations [and discussions](https://gitter.im/angular/angular?at=57a9f718bc688a1339093273) about how to obtain control over the update mechanism. Now I ran into another surprise. Therefore I propose one of the two
1. Code changes in order to have a more consistent and/or intuitive connection between `ngDoCheck`, `ngOnChanges` and `ChangeDetectorRef`. Maybe some interception makes no sense
2. Documentation changes, i.e. some more clarification added about 
   - `markForCheck` vs. `dectecChanges`
   - How all those `*check*` play together

Please note that _works as intended_ is one thing, _works as intuitively expected_ another. E.g. it is completely surprising that declarig `changeDetection: ChangeDetectionStrategy.OnPush` overrules `changeDetectorRef.detach()` - at least that is what I see in my experiments. I find neither the docs nor [the closing comment of #7054](https://github.com/angular/angular/issues/7054#issuecomment-227306686) clear and helpful on the topic of fine grained control on updates and renderings.

**Please tell us about your environment:**
- **Angular version:** current =  https://npmcdn.com/@angular
- **Browser:** all
- **Language:** TypeScript

**Comment by daixtrose** [comment]
Since a plunkr might get lost, I post the demo code for completeness here:

```typescript
//our root app component
import {Component, ChangeDetectionStrategy, Input, NgZone, DoCheck, OnChanges, ChangeDetectorRef} from '@angular/core';
import {Observable, Subject} from 'rxjs/Rx';

@Component({
  selector: 'on-push-update',
  changeDetection: ChangeDetectionStrategy.OnPush,
  template: `
  <h3>{{speed}} speed:</h3> 
  
  <table>
  <tr>
    <th></th>
    <th colspan="3">Number of calls to</th>
  </tr>
  <tr>
    <th>Value</th>
    <th>Change Detections</th> 
    <th>ngOnChanges</th>
    <th>ngDoCheck</th>
  </tr>
  <tr>
    <td>{{value}}</td>
    <td><mark>{{changeDetectiondCount}}</mark></td> 
    <td><mark>{{onChangesCount }}</mark></td>
    <td>{{ doCheckCount }}</td>
  </tr>
</table>
`
})
export class OnPushUpdater implements DoCheck, OnChanges {
  @Input('value') value: number = 0;
  @Input('speed') speed: string;
  
  _changeDetectionCount: number = 0;
  get changeDetectiondCount(): number {
    return ++this._changeDetectionCount;
  }
  
  _onChangesCount: number = 0;
  ngOnChanges(changes:any):void {
      ++ this._onChangesCount;
  }
  
  get onChangesCount(): number {
    return this._onChangesCount;
  }
  
  ngDoCheck() {
    ++this._doCheckCount;
  }
  
  _doCheckCount: number = 0;
  get doCheckCount(): number {
    return this._doCheckCount;
  }
}


@Component({
  selector: 'has-decoupled-change-detection',
  // DOES NOT! USE changeDetection: ChangeDetectionStrategy.OnPush,
  // BECAUSE THAT DOES NOT PLAY WELL WITH changeDetectorRef.detach()
  template: `
  <h3>{{speed}} speed:</h3>   <table>
  <tr>
    <th></th>
    <th colspan="3">Number of calls to</th>
  </tr>
  <tr>
    <th>Value</th>
    <th>Change Detections</th> 
    <th>ngOnChanges</th>
    <th>ngDoCheck</th>
  </tr>
  <tr>
    <td>{{value}}</td>
    <td><mark>{{changeDetectiondCount}}</mark></td> 
    <td><mark>{{onChangesCount }}</mark></td>
    <td>{{ doCheckCount }}</td>
  </tr>
</table>`
})
export class HasDecoupledChangeDetection implements DoCheck, OnChanges {
  @Input('value') value: number = 0;
  @Input('speed') speed: string;
  
  constructor(private changeDetectorRef: ChangeDetectorRef) {
    changeDetectorRef.detach();
    // cannot call this for an initial draw, so first draw happens later
    // this.changeDetectorRef.detectChanges();
    setInterval(() => {
        console.log("--> Triggering detectChanges" + new Date());
        this.changeDetectorRef.detectChanges();
    }, 3000);
  }
  
  _changeDetectionCount: number = 0;
  get changeDetectiondCount(): number {
    return ++this._changeDetectionCount;
  }
  
  _onChangesCount: number = 0;
  ngOnChanges(changes:any):void {
      ++ this._onChangesCount;
  }
  
  get onChangesCount(): number {
    return this._onChangesCount;
  }
  
  ngDoCheck() {
    ++this._doCheckCount;
  }
  
  _doCheckCount: number = 0;
  get doCheckCount(): number {
    return this._doCheckCount;
  }
}


@Component({
  selector: 'my-app',
  providers: [],
  template: `
<h1>Using regular update mechanisms</h1>  
<on-push-update speed="High" [value]="highSpeed | async"></on-push-update>
<on-push-update speed="Low" [value]="lowSpeed | async"></on-push-update>
<h1>Detached change detector (fires every 3 seconds)</h1>
<has-decoupled-change-detection speed="High" [value]="highSpeed | async"></has-decoupled-change-detection>
<has-decoupled-change-detection speed="Low" [value]="highSpeed | async"></has-decoupled-change-detection>
  `,
  directives: [OnPushUpdater, HasDecoupledChangeDetection]
})
export class App {
  data: Subject<number> = new Subject();
  highSpeed: Subject<number> = new Subject();
  lowSpeed: Subject<number> = new Subject();
  
  constructor() {
    this.data.subscribe(v => {
      if (v % 5 == 0) {
        this.highSpeed.next(v);
      }
      if (v % 20 == 0) {
        this.lowSpeed.next(v);
      }
    });
    let i: number = 0;
    setInterval(() => {
      this.data.next(i++);
    }, 100);
  }
}
```

**Comment by daixtrose** [comment]
It seems that `ngOnChanges` gets only called when the window is active (visible?), but change detection occurs always. This too is not easily extracted from the docs.

**Comment by pkozlowski-opensource** [comment]
@daixtrose could you please confirm that this is still the case in 2.0.3 and add a plunker if this happens to you? You haven't added plunker (one linked is a default one) and you've got quite a bunch of code here.

Please provide a minimal reproduce scenario in plunker. Please check out [our submission guidelines](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-submitting-an-issue) to understand why we can't act on issues that are lacking important information.

**Comment by DzmitryShylovich** [comment]
@pkozlowski-opensource no repro so it can be closed I think

**Comment by chuckjaz** [comment]
Thanks for reporting this issue. This issue is now obsolete due to changes in the recent releases. Please update to the most recent Angular version.

If the problem still persists, please file a new issue and ensure you provide the version of Angular affected and include the steps to reproduce the problem when filling out the issue template.
